### ISSUE TITLE
tf.contrib.learn.estimator each call to predict() reloads the model

### ISSUE THREAD

**Issue by ashern** [tensorflow/tensorflow#4648]
### What related GitHub issues or StackOverflow threads have you found by searching the web for your problem?
None

### Environment info
Operating System:
OSX & Ubuntu 14.04

### If possible, provide a minimal reproducible example - Explanation

Each call to estimator.predict() completely reloads the model, which is very slow. Is there any mechanism to maintain the model in GPU/memory for consecutive calls? 

I understand that large arrays can be passed to the function for simultaneous predictions, but my use case is for slowly arriving data which must be processed quickly. Is there a mechanism for decreasing the 'start-up' time of this function?

### What other attempted solutions have you tried?

Looked through API docs & source code for alternative solutions

###########################

**Comment by ashern** [comment]
I discussed this issue briefly with Martin Wicke on the mailing list: https://groups.google.com/a/tensorflow.org/forum/#!topic/discuss/rOP4VKcfphg .

It seems that this use case is not supported OOB. 

The ability to persistently load a model into memory for multiple predict calls would be a very useful enhancement.

**Comment by ahmed-18** [comment]
Hi @ashern 
Did you find a workaround? Anything that prevents reloading the model on each predict call?

**Comment by ahmed-18** [comment]
I have modified estimator.py (Added my own predict function in BaseEstimator class). It's working for me. It's available [here](https://github.com/ahmed-18/tensorflow/tree/master/tf.contrib.learn) for anyone's use.

**Comment by marcsto** [comment]
In case someone else runs into this, predict can take a generator as a parameter. 

Here's a class that wraps an estimator with a generator. This means you can make multiple calls to predict without reloading the graph.

https://github.com/marcsto/rl/blob/master/src/fast_predict.py

**Comment by ashern** [comment]
Thanks Marc - that is exactly the pattern I was describing.

**Comment by AVCarreiro** [comment]
Hi Marc, with the deprecation of sklearn compatibility in recent versions, how do you see the changes? I tried wrapping the estimator with SKCompat to keep using x as argument but then I have problems down the line. It seems to me that keeping input_fn open as a generator would make perfect sense, especially for big data (not relying only on parsing TFrecords and such)

**Comment by secsilm** [comment]
Hi guys, I have trained my cnn model with `tf.estimator` and `tf.data` API. Model's input is an one-shot iterator. The relative model files(ckpt, meta, index) have been saved in `model_dir`. **Now I want to predict one or more images' label using the trained model.**

Any instructions for this?

**Comment by secsilm** [comment]
@j3ven7 I wrote a example use `tf.estimator`, whick separate the train and predict process in different python file. You can see [secsilm/understaing-datasets-estimators-tfrecords](https://github.com/secsilm/understaing-datasets-estimators-tfrecords). There is a chinese README.md, but the two main python files are [cifar10_estimator_dataset.py](https://github.com/secsilm/understaing-datasets-estimators-tfrecords/blob/master/cifar10_estimator_dataset.py) for train and [cifar10_estimator_dataset_predict.py](https://github.com/secsilm/understaing-datasets-estimators-tfrecords/blob/master/cifar10_estimator_dataset_predict.py) for predict. 

Hope it helps you.

**Comment by j3ven7** [comment]
@secsilm Does the code you wrote address the issue regarding the model re-loading on each predict? Or is this strictly a tutorial on how to train and predict?

**Comment by secsilm** [comment]
@j3ven7 I don't know whether it's a right solution or not. I combine 10 images to an numpy array and pass it to the predict function in my code. But the point is, I think tensorflow will reload the model in different run. Might not solve your problem? For reference only. 😄

**Comment by marcsto** [comment]
I updated fast_predict2.py to support tensorflow 1.4 and later. I hope it helps!
https://github.com/marcsto/rl/blob/master/src/fast_predict2.py

**Comment by thak123** [comment]
@marcsto  fast_predict test file is still old, can u please fix it.Any idea on how can we pass pandas df to it

**Comment by mikeoconnor0308** [comment]
Thanks to @marcsto, I got this working for a DNNClassifier that's using several numeric feature columns. I found it a bit tricky to figure out exactly how to set up the generator, so there's a gist here: 

https://gist.github.com/mikeoconnor0308/521ae2eb1555edc6550014ce0500e6a2

**Comment by schulz0r** [comment]
I tried to find a shorter solution and came out with a generator function which is fed using the build-in send() function:
```Python
def inference_generator(estimator):
    predictor_input = yield # wait here until send()
    predict_input_fn = lambda: predictor_input # input func just returns a single numpy array
    predictor = estimator.predict(input_fn=predict_input_fn, predict_keys=['features'])
    yield next(predictor) # yield for the first time
    while True:
        predictor_input = yield
        yield next(predictor)
```

So, here is how to use it:
```Python
gen = inference_generator(my_estimator)
next(gen) # start generator
# generator with an numpy array (do it as often as you want to)
result = gen.send(np.ones([1, 16, 16, 3], dtype=np.float32))
```

**Comment by SaulAryehKohn** [comment]
@mikeoconnor0308 @marcsto thanks, this is exactly the issue I've been facing!

**Comment by wenbotse** [comment]
another way can work around this problem,  you can define a "input_fn" with a dataset , and you can use dataset.from_generator(generator, ****),  in the definition of 'generator' you can use yield after read from outside queue , such as Kafka。and estimator predict also return an generator. You can get the result from the predict result generator supply to the caller

**Comment by dakshvar22** [comment]
@marcsto I tried using your script and wanted the model to use a batch size of more than 1 and hence changed - `.batch(1)` to `.batch(batch_size)`.
For some reason the model does not receive new input on subsequent calls and the outputs from the model are completely destroyed on any call after the first call. If I send the same input in subsequent calls as the first call, then the output is fine. I am guessing there is some kind of flushing out of data that I need to do but I am not sure why that would be needed. Any pointers?
